unittests/producer_schedule_tests.cpp

```diff
@@ -503,4 +503,224 @@ BOOST_AUTO_TEST_CASE( producer_watermark_test ) try {
 
 } FC_LOG_AND_RETHROW()
 
+BOOST_FIXTURE_TEST_CASE( producer_one_of_n_test, TESTER ) try {
+   create_accounts( {N(alice),N(bob)} );
+   while (control->head_block_num() < 3) {
+      produce_block();
+   }
+
+   auto compare_schedules = [&]( const vector<producer_authority>& a, const producer_authority_schedule& b ) {
+      return std::equal( a.begin(), a.end(), b.producers.begin(), b.producers.end() );
+   };
+
+   vector<producer_authority> sch1 = {
+                                 producer_authority{N(alice), block_signing_authority_v0{1, {{get_public_key(N(alice), "bs1"), 1}, {get_public_key(N(alice), "bs2"), 1}}}},
+                                 producer_authority{N(bob),   block_signing_authority_v0{1, {{get_public_key(N(bob),   "bs1"), 1}, {get_public_key(N(bob),   "bs2"), 1}}}}
+                               };
+
+   auto res = set_producer_schedule( sch1 );
+   block_signing_private_keys.emplace(get_public_key(N(alice), "bs1"), get_private_key(N(alice), "bs1"));
+   block_signing_private_keys.emplace(get_public_key(N(bob),   "bs1"), get_private_key(N(bob),   "bs1"));
+
+   //wdump((fc::json::to_pretty_string(res)));
+   wlog("set producer schedule to [alice,bob]");
+   BOOST_REQUIRE_EQUAL( true, control->proposed_producers().valid() );
+   BOOST_CHECK_EQUAL( true, compare_schedules( sch1, *control->proposed_producers() ) );
+   BOOST_CHECK_EQUAL( control->pending_producers().version, 0u );
+   produce_block(); // Starts new block which promotes the proposed schedule to pending
+   BOOST_CHECK_EQUAL( control->pending_producers().version, 1u );
+   BOOST_CHECK_EQUAL( true, compare_schedules( sch1, control->pending_producers() ) );
+   BOOST_CHECK_EQUAL( control->active_producers().version, 0u );
+   produce_block();
+   produce_block(); // Starts new block which promotes the pending schedule to active
+   BOOST_CHECK_EQUAL( control->active_producers().version, 1u );
+   BOOST_CHECK_EQUAL( true, compare_schedules( sch1, control->active_producers() ) );
+   produce_blocks(18);
```

Comment:
For example this could be replaced by `produce_empty_blocks_until( *this, N(bob), N(alice) );`.

You could even put a limit on the number of blocks to produce to avoid getting into an infinite loop if something went wrong earlier with setting up the producer schedule: `produce_empty_blocks_until( *this, N(bob), N(alice), 36 );`. In that case, one of the `BOOST_REQUIRE_EQUAL` assertions would fail and the test would terminate.

Or even better, since the `produce_empty_blocks_until` returns true iff the desired condition is met, you can avoid the next two `BOOST_REQUIRE_EQUAL` lines as well and just do the following:
```cpp
BOOST_REQUIRE( produce_empty_blocks_until( *this, N(bob), N(alice), 36 ) );
```

Reply:
I ended up deciding that I did not need to re-test the schedule promotion logic in all these cases.  What those tests should show is that 1-of-n and m-of-n wtmsig block signing authorities are valid.   Most of the logic was replace by a new version of `produce_empty_blocks_until` that will produce blocks until all producers in a `std::set` have had a turn (or until a threshold of blocks is reached).  This makes those tests a lot shorter.  It's just setting a schedule and guaranteeing that in the first N blocks those producers can produce valid blocks.